test/integration/framework/etcd.go

```diff
@@ -17,93 +17,113 @@ limitations under the License.
 package framework
 
 import (
+	"context"
 	"fmt"
-	"hash/adler32"
-	"io"
 	"io/ioutil"
 	"math/rand"
+	"net"
 	"os"
 	"os/exec"
 	"path/filepath"
-	"sync"
+	"strings"
 
 	"github.com/golang/glog"
 
 	"k8s.io/kubernetes/pkg/util/env"
 )
 
-var (
-	etcdSetup sync.Once
-	etcdURL   = ""
-)
+var etcdURL = ""
 
-func setupETCD() {
-	etcdSetup.Do(func() {
-		if os.Getenv("RUNFILES_DIR") == "" {
-			etcdURL = env.GetEnvAsStringOrFallback("KUBE_INTEGRATION_ETCD_URL", "http://127.0.0.1:2379")
-			return
-		}
-		etcdPath := filepath.Join(os.Getenv("RUNFILES_DIR"), "com_coreos_etcd/etcd")
-		// give every test the same random port each run
-		etcdPort := 20000 + rand.New(rand.NewSource(int64(adler32.Checksum([]byte(os.Args[0]))))).Intn(5000)
-		etcdURL = fmt.Sprintf("http://127.0.0.1:%d", etcdPort)
+const installEtcd = `
+Cannot find etcd, cannot run integration tests
+Please see https://github.com/kubernetes/community/blob/master/contributors/devel/testing.md#install-etcd-dependency for instructions.
 
-		info, err := os.Stat(etcdPath)
-		if err != nil {
-			glog.Fatalf("Unable to stat etcd: %v", err)
-		}
-		if info.IsDir() {
-			glog.Fatalf("Did not expect %q to be a directory", etcdPath)
-		}
+You can use 'hack/install-etcd.sh' to install a copy in third_party/.
 
-		etcdDataDir, err := ioutil.TempDir(os.TempDir(), "integration_test_etcd_data")
-		if err != nil {
-			glog.Fatalf("Unable to make temp etcd data dir: %v", err)
-		}
-		glog.Infof("storing etcd data in: %v", etcdDataDir)
-
-		etcdCmd := exec.Command(
-			etcdPath,
-			"--data-dir",
-			etcdDataDir,
-			"--listen-client-urls",
-			GetEtcdURL(),
-			"--advertise-client-urls",
-			GetEtcdURL(),
-			"--listen-peer-urls",
-			"http://127.0.0.1:0",
-		)
-
-		stdout, err := etcdCmd.StdoutPipe()
-		if err != nil {
-			glog.Fatalf("Failed to run etcd: %v", err)
-		}
-		stderr, err := etcdCmd.StderrPipe()
+`
+
+// getEtcdPath returns a path to an etcd executable.
+func getEtcdPath() (string, error) {
+	bazelPath := filepath.Join(os.Getenv("RUNFILES_DIR"), "com_coreos_etcd/etcd")
+	p, err := exec.LookPath(bazelPath)
+	if err == nil {
+		return p, nil
+	}
+	return exec.LookPath("etcd")
+}
+
+// startEtcd executes an etcd instance. The returned function will signal the
+// etcd process and wait for it to exit.
+func startEtcd() (func(), error) {
+	etcdURL = env.GetEnvAsStringOrFallback("KUBE_INTEGRATION_ETCD_URL", "http://127.0.0.1:2379")
+	conn, err := net.Dial("tcp", strings.TrimPrefix(etcdURL, "http://"))
+	if err == nil {
+		glog.Infof("etcd already running at %s", etcdURL)
+		conn.Close()
+		return func() {}, nil
+	}
+	glog.V(1).Infof("could not connect to etcd: %v", err)
+
+	// TODO: Check for valid etcd version.
+	etcdPath, err := getEtcdPath()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, installEtcd)
+		return nil, fmt.Errorf("could not find etcd in PATH: %v", err)
+	}
+	// TODO: Pick an unused port.
```

Comment:
this part worries me a little bit. since bazel may run multiple integration tests in parallel, we want to keep the etcd instances separate, hence previously it used a constant "random" port. without any means to ensure this port is unused, this seems like it will cause test flakes.

Reply:
I can put it back to the way it was, but IMO the previous behavior encodes assumptions about argv[0] that are not obvious. How do I know if argv[0] == 'make' or 'bazel' or 'go'?

I was thinking that the classic, "get a free port from the kernel and then release it" method would be more suitable.

Or I can put the old behavior back.

Reply:
I think `argv[0]` would be the name of the go test that was compiled, e.g. under bazel it might be something like `bazel-bin/test/integration/auth/linux_amd64_stripped/go_default_test`.

I'm not sure whether this does the right thing under just `go test` though.

Reply:
though yes, picking an actual free port would be best. the current method of using the test name doesn't help if we run 10 instances of the same test in parallel, e.g. to shake out flakes.

Reply:
Done. Thanks for the feedback!